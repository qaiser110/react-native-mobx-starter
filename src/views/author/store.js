import { types, process, getEnv } from 'mobx-state-tree'
import api from './__mocks__/api'

export const Author = types.model('Author', {
  id: types.identifier(types.number),
  name: types.string,
  gender: types.enumeration('Gender', ['Male', 'Female']),
})

export const AuthorStore = types
  .model('AuthorStore', {
    isLoading: true,
    authors: types.map(Author),
  })
  .views(self => ({
    get api() {
      return getEnv(self) && getEnv(self).api ? getEnv(self).api : api
    },
  }))
  .actions(self => {
    function markLoading(loading) {
      self.isLoading = loading
    }

    function updateAuthors(json) {
      json.forEach(authorJson => {
        self.authors.put(authorJson)
      })
    }

    const loadAuthors = process(function* loadAuthors() {
      try {
        const json = yield self.api.fetchAuthors('/authors')
        updateAuthors(json)
        markLoading(false)
      } catch (err) {
        console.error('Failed to load authors ', err)
      }
    })

    return {
      loadAuthors,
    }
  })
