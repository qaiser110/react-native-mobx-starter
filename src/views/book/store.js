import { types, process, getEnv } from 'mobx-state-tree'

export const Book = types.model('Book', {
  id: types.identifier(),
  name: types.string,
  author: types.string,
  genre: types.string,
  inStock: types.boolean,
})

export const BookStore = types
  .model('BookStore', {
    isLoading: true,
    books: types.map(Book),
  })
  .views(self => ({
    get api() {
      return getEnv(self).api
    },
    booksByGenre(genre) {
      return self.books.values().filter(t => t.genre === genre)
    },
    get sortedAvailableBooks() {
      return sortBooks(self.books.values())
    },
  }))
  .actions(self => {
    function markLoading(loading) {
      self.isLoading = loading
    }

    function updateBooks(json) {
      json.forEach(bookJson => {
        self.books.put(bookJson)
      })
    }

    const loadBooks = process(function* loadBooks() {
      try {
        const json = yield self.api.fetchBooks('/books')
        updateBooks(json)
        markLoading(false)
      } catch (err) {
        console.error('Failed to load books ', err)
      }
    })

    return {
      loadBooks,
    }
  })

function sortBooks(books) {
  return books
    .filter(b => b.inStock)
    .sort((a, b) => (a.name > b.name ? 1 : a.name === b.name ? 0 : -1))
}
