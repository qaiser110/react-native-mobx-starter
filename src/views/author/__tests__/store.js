import * as fs from 'fs'
import { when } from 'mobx'
import { getSnapshot } from 'mobx-state-tree'
import { AuthorStore } from '../store'

const api = {
  fetchAuthors() {
    return Promise.resolve(
      JSON.parse(fs.readFileSync(`${__dirname}/../__mocks__/authors.json`))
    )
  },
}

it('store.authors.values() data is correct', done => {
  const store = AuthorStore.create({ authors: {} }, { api })
  store.loadAuthors()
  when(
    () => store.isLoading === false,
    () => {
      expect(store.authors.values()).toMatchSnapshot()
      done()
    }
  )
})

it('author getSnapshot(store) data is correct', done => {
  const store = AuthorStore.create({ authors: {} }, { api })
  store.loadAuthors()
  when(
    () => store.isLoading === false,
    () => {
      expect(getSnapshot(store)).toMatchSnapshot()
      done()
    }
  )
})
