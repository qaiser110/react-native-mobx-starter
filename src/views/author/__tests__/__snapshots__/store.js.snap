// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`author getSnapshot(store) data is correct 1`] = `
Object {
  "authors": Object {
    "371": Object {
      "gender": "Female",
      "id": 371,
      "name": "Rebecca Skloot",
    },
    "372": Object {
      "gender": "Male",
      "id": 372,
      "name": "F. Scott Fitzgerald",
    },
    "373": Object {
      "gender": "Male",
      "id": 373,
      "name": "Rick Riordan",
    },
    "374": Object {
      "gender": "Male",
      "id": 374,
      "name": "Malcolm Gladwell",
    },
  },
  "isLoading": false,
}
`;

exports[`store.authors.values() data is correct 1`] = `
Array [
  Object {
    "gender": "Female",
    "id": 371,
    "name": "Rebecca Skloot",
  },
  Object {
    "gender": "Male",
    "id": 372,
    "name": "F. Scott Fitzgerald",
  },
  Object {
    "gender": "Male",
    "id": 373,
    "name": "Rick Riordan",
  },
  Object {
    "gender": "Male",
    "id": 374,
    "name": "Malcolm Gladwell",
  },
]
`;
